### DISCUSSION TITLE
Is there a way to import configuration map values without removing existing ones?

### DISCUSSION THREAD

**Discussion by jerchap** [nextgenhealthcare/connect#5669] [Q&A]
I would like to be able to import subsets of values into the configuration map without removing existing items.  both the CLI and the Mirth admin UI seem to remove items that are not in a config map file.

**Comment by jonbartels** [comment]
The Mirth Client has several places where it operates on entire sets of objects instead of discrete rows. Another example is that when you change channel group assignments, Mirth does not just touch one channel, it rebuilds the WHOLE list of channel groups and channel assignments to those groups. 

I think you have a few options:
1. You can manipulate the configuration.properties file
2. You can change `mirth.properties` `configurationmap.location` to `database` and MC will store the config map in the database. Then you can manipulate it using DB queries. 

I am attaching a channel I wrote which will:
1. Read the `configurationmap.location` property. If it is "database" then the channel undeploys itself
2. If it is "file" then it will read each configmap property and write it to the SAME location Mirth uses in its own DB
3. Then you can stop mirth, change mirth.properties to make configmap use the DB
4. Start mirth and all your properties will be the same but they're stored in the DB

For your use case, now you can do DB updates or you can see how I use the ConfigurationController calls and upsert into the config to make changes or merge exports. 

You can see what I am describing on line 39 in the source connector - see how the WHOLE map is saved and not just properties one by one? Thats the fundamental problem you're experiencing. 


[Synchronize configMap from file to db.txt](https://github.com/nextgenhealthcare/connect/files/10701684/Synchronize.configMap.from.file.to.db.txt)

**Reply by pacmano1** [reply]
You can just read the current config map via API, and post back your updates (which would be prior list + your changes), no?

**Reply by rutledj** [reply]
So I guess I need a little more information on how this works. 

Assume I have a mysql.properties file with these contents:
```
vETL_PROD_PD05URL=jdbc:jtds:sqlserver://ACUEHRSQLPD16.ABC.COM:27702;
vUserName=svcMirthUserNameProd
vPassword=abc123
vETL_DEV_DV05URL=jdbc:jtds:sqlserver://ACUEHRSQLDV16.ABC.COM:27702;
vUserName=svcMirthUserNameDev
vPassword=abc456
```
If I add configurationmap.location = file to my mirth.properties file (it currently isn't in there),

The channel will save this data to the same db that mirth uses for it's own data purposes.

My question is, how do you then access that information from another channel?  I want to connect to vETL_PROD_PD05URL. How do I get the URL and credentials?

**Reply by jonbartels** [reply]
In your channel you can just reference `${vETL_PROD_PD05URL}` in the GUI config. Or if you are using Javascript you can call `$('vETL_PROD_PD05URL')`. That will read from the maps in the order documented here - https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_Variable_Maps_java_connect_ug.html

**Reply by pacmano1** [reply]
You should probably be using a code template for DB calls.  like https://github.com/pacmano1/Mirth-Snippets/blob/main/executeSharedDBStatement.js

That is one of a few out there, note the bottom comments.

In your DB readers/writers you would use JS.

For example in a deploy script:

```JS
var limit = "limit 10"
var sql = <>
select * from table
{limit}
</>.toString()
$gc('SQLSource',sql)
```
on your DB reader on the source, set to type javascript and:
```JS
var result = executeSharedDBStatement($gc('SQLSource'),true)
return result
```

An important point here is you only ever need to reference the DB connect values once.

**Comment by rutledj** [comment]
Ok. Sorry, we really haven't used the configuration map any at work, instead saving all our data to a file we read into the globalMap. I just watched a video about it, and it makes much more sense now. We currently do use code templates (almost exactly like yours) to get our db connections.

Going back to the original post by @jonbartels about that channel that saves the config values to the db, I'm assuming that is a type of backup precaution? Not really understanding how you get those values back if for some reason they are needed.
Also, if the config file got corrupted in some way, wouldn't that channel write the bad data to the db?

**Reply by pacmano1** [reply]
The config map by default saves to disk on the mirth server.  Not the greatest place for it, the db is better.   Sure, you can corrupt it like anything else.

But if it were corrupt I would expect the db connection to fail not write bad data.

**Reply by jonbartels** [reply]
The manual might explain it better than I can! :P 

What the config map is - https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_Configuration_MapConfiguration_Map_connect_ug.html

How the configmap is stored, look at `configurationmap.location` : https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_The_mirth_properties_File_connect_ug.html
